File: python/src/heaton_life/fractal/perturbation.py
"""Perturbation + rebasing deep-zoom engine (spec/deep-zoom.md).

One high-precision reference orbit (core.bignum); every pixel iterates its small
deviation delta in plain float64. Rebasing (Zhuoran 2021): whenever the full value
|Z[m] + delta| drops below |delta|, restart against the beginning of an orbit whose
first sample is 0 (delta <- Z[m] + delta, m <- 0). For Mandelbrot and Burning Ship
that is the reference itself (Z0 = 0); a Julia reference starts at the viewport
center, so Julia rebases onto the critical orbit (W0 = 0, same c) instead. One
reference serves the whole frame; no glitch detection passes. The reference index
is clamped to the last sample of whichever orbit the pixel follows, which is only
reachable when that orbit escaped — those pixels escape immediately.
"""

from __future__ import annotations

import numpy as np
from numpy.typing import NDArray

ComplexArray = NDArray[np.complex128]
FloatArray = NDArray[np.float64]
IntArray = NDArray[np.int32]


def perturb_z2(
    orbit: ComplexArray,
    delta0: ComplexArray,
    delta_c: ComplexArray,
    max_iter: int,
    escape_radius: float,
    rebase_orbit: ComplexArray | None = None,
) -> tuple[IntArray, ComplexArray]:
    """Perturbation for z^2 + c maps (Mandelbrot: delta0=0; Julia: delta_c=0).

    ``rebase_orbit`` is the orbit a rebased pixel restarts on; it must begin at 0.
    None means the reference itself (Mandelbrot, whose reference starts at 0).
    Julia passes its critical orbit. Same (counts, final_z) contract as
    engine.escape_time.
    """
    counts, final, _ = _perturb_z2(
        orbit, delta0, delta_c, max_iter, escape_radius, rebase_orbit, None
    )
    return counts, final


def perturb_z2_distance(
    orbit: ComplexArray,
    delta0: ComplexArray,
    delta_c: ComplexArray,
    max_iter: int,
    escape_radius: float,
    d0: tuple[float, float],
    add: float | None,
    rebase_orbit: ComplexArray | None = None,
) -> tuple[IntArray, ComplexArray, tuple[FloatArray, FloatArray]]:
    """perturb_z2 that also carries the derivative of the full z for a distance
    estimate (spec/fractals.md "Distance estimate"): updated from the z the previous
    iteration reconstructed -- fl(Z_0 + delta0) before the first -- and left alone by a
    rebase, which changes delta, m and the orbit but not z. ``d0`` and ``add`` as in
    engine.escape_time_distance. Counts and final z are exactly perturb_z2's."""
    counts, final, derivative = _perturb_z2(
        orbit, delta0, delta_c, max_iter, escape_radius, rebase_orbit, (d0, add)
    )
    assert derivative is not None
    return counts, final, derivative


def _perturb_z2(
    orbit: ComplexArray,
    delta0: ComplexArray,
    delta_c: ComplexArray,
    max_iter: int,
    escape_radius: float,
    rebase_orbit: ComplexArray | None,
    derivative: tuple[tuple[float, float], float | None] | None,
) -> tuple[IntArray, ComplexArray, tuple[FloatArray, FloatArray] | None]:
    # Both orbits live in one array so a pixel's reference sample stays a single
    # fancy-indexed gather: m is an absolute index, `last` the end of the orbit
    # that pixel currently follows. With no separate rebase orbit the values are
    # exactly the historical single-orbit loop's.
    if rebase_orbit is None:
        refs = orbit
        rebase_start = 0
    else:
        refs = np.concatenate([orbit, rebase_orbit])
        rebase_start = len(orbit)
    rebase_last = len(refs) - 1
    n = delta0.size
    counts = np.full(n, -1, dtype=np.int32)
    final = np.zeros(n, dtype=np.complex128)
    dz = delta0.copy()
    dc = delta_c.copy()
    m = np.zeros(n, dtype=np.int64)
    last = np.full(n, len(orbit) - 1, dtype=np.int64)
    idx = np.arange(n)
    r2 = escape_radius * escape_radius
    dr: FloatArray | None = None
    di: FloatArray | None = None
    final_dr = final_di = np.zeros(0)
    add: float | None = None
    if derivative is not None:
        (d0r, d0i), add = derivative
        dr = np.full(n, d0r, dtype=np.float64)
        di = np.full(n, d0i, dtype=np.float64)
        final_dr = np.zeros(n, dtype=np.float64)
        final_di = np.zeros(n, dtype=np.float64)
        z = refs[m] + dz  # the pre-square z of iteration 1: fl(Z_0 + delta0)
    for it in range(1, max_iter + 1):
        if dr is not None and di is not None:
            zr = z.real
            zi = z.imag
            tr = 2.0 * (zr * dr - zi * di)
            if add is not None:
                tr = tr + add
            ti = 2.0 * (zr * di + zi * dr)
            dr, di = tr, ti
        dz = (2.0 * refs[m] + dz) * dz + dc
        m = np.minimum(m + 1, last)
        z = refs[m] + dz
        zabs2 = z.real * z.real + z.imag * z.imag
        escaped = zabs2 > r2
        if escaped.any():
            hits = idx[escaped]
            counts[hits] = it
            final[hits] = z[escaped]
            keep = ~escaped
            dz, dc, m, last, idx, z, zabs2 = (
                dz[keep],
                dc[keep],
                m[keep],
                last[keep],
                idx[keep],
                z[keep],
                zabs2[keep],
            )
            if dr is not None and di is not None:
                final_dr[hits] = dr[escaped]
                final_di[hits] = di[escaped]
                dr, di = dr[keep], di[keep]
            if idx.size == 0:
                break
        rebase = zabs2 < (dz.real * dz.real + dz.imag * dz.imag)
        if rebase.any():
            # z, and so the derivative, are unchanged: only delta, m and the orbit move
            dz[rebase] = z[rebase]
            m[rebase] = rebase_start
            last[rebase] = rebase_last
    return counts, final, (final_dr, final_di) if derivative is not None else None


def perturb_burning_ship(
    orbit: ComplexArray,
    delta_c: ComplexArray,
    max_iter: int,
    escape_radius: float,
) -> tuple[IntArray, ComplexArray]:
    """Component-form perturbation for the Burning Ship, using stable diffabs."""
    ref_x = orbit.real.copy()
    ref_y = orbit.imag.copy()
    n = delta_c.size
    counts = np.full(n, -1, dtype=np.int32)
    final = np.zeros(n, dtype=np.complex128)
    dx = np.zeros(n, dtype=np.float64)
    dy = np.zeros(n, dtype=np.float64)
    dcx = delta_c.real.copy()
    dcy = delta_c.imag.copy()
    m = np.zeros(n, dtype=np.int64)
    idx = np.arange(n)
    last = len(orbit) - 1
    r2 = escape_radius * escape_radius
    for it in range(1, max_iter + 1):
        x_ref = ref_x[m]
        y_ref = ref_y[m]
        a = _diffabs(x_ref, dx)
        b = _diffabs(y_ref, dy)
        new_dx = (2.0 * x_ref + dx) * dx - (2.0 * y_ref + dy) * dy + dcx
        new_dy = 2.0 * (np.abs(x_ref) * b + np.abs(y_ref) * a + a * b) + dcy
        dx, dy = new_dx, new_dy
        m = np.minimum(m + 1, last)
        zx = ref_x[m] + dx
        zy = ref_y[m] + dy
        zabs2 = zx * zx + zy * zy
        escaped = zabs2 > r2
        if escaped.any():
            hits = idx[escaped]
            counts[hits] = it
            final[hits] = zx[escaped] + 1j * zy[escaped]
            keep = ~escaped
            dx, dy, dcx, dcy, m, idx, zx, zy, zabs2 = (
                dx[keep],
                dy[keep],
                dcx[keep],
                dcy[keep],
                m[keep],
                idx[keep],
                zx[keep],
                zy[keep],
                zabs2[keep],
            )
            if idx.size == 0:
                break
        rebase = zabs2 < (dx * dx + dy * dy)
        if rebase.any():
            dx[rebase] = zx[rebase]
            dy[rebase] = zy[rebase]
            m[rebase] = 0
    return counts, final


def _diffabs(ref: FloatArray, delta: FloatArray) -> FloatArray:
    """|ref + delta| - |ref|, computed without cancellation (case analysis)."""
    total = ref + delta
    result: FloatArray = np.where(
        ref >= 0.0,
        np.where(total >= 0.0, delta, -(2.0 * ref + delta)),
        np.where(total <= 0.0, -delta, 2.0 * ref + delta),
    )
    return result

File: dotnet/src/HeatonLife.Core/Perturbation.cs
using System;

namespace HeatonLife
{
    /// <summary>
    /// Perturbation + rebasing deep-zoom engine (spec/deep-zoom.md). One high-precision
    /// reference orbit (<see cref="ReferenceOrbit"/>, or a stored vector) and every
    /// pixel iterates its small deviation delta in plain float64. Rebasing (Zhuoran
    /// 2021): whenever the full value |Z[m] + delta| drops below |delta|, restart
    /// against the beginning of an orbit whose first sample is 0 — the reference
    /// itself for Mandelbrot and Burning Ship, the critical orbit for Julia (whose
    /// reference starts at the viewport center). One reference serves the whole
    /// frame; no glitch detection passes.
    /// </summary>
    public static class Perturbation
    {
        /// <summary>
        /// Perturbation for z^2 + c maps whose reference starts at 0 (Mandelbrot:
        /// delta0 = 0), so a rebased pixel restarts on the reference itself.
        /// Returns the 1-based escape iteration for one pixel, or -1.
        /// </summary>
        public static int PerturbZ2(
            double[] orbitRe,
            double[] orbitIm,
            double dz0Re,
            double dz0Im,
            double dcRe,
            double dcIm,
            int maxIter,
            double escapeRadius,
            out double finalRe,
            out double finalIm)
            => PerturbZ2(
                orbitRe, orbitIm, orbitRe, orbitIm, dz0Re, dz0Im, dcRe, dcIm,
                maxIter, escapeRadius, out finalRe, out finalIm);

        /// <summary>
        /// Perturbation for z^2 + c maps with a separate rebase orbit, which must
        /// start at 0 (Julia: deltaC = 0, the reference starts at the viewport
        /// center, and <paramref name="rebaseRe"/>/<paramref name="rebaseIm"/> is the
        /// critical orbit under the same c). A pixel follows the reference until its
        /// first rebase and the rebase orbit from then on (spec/deep-zoom.md
        /// "Rebasing"). Returns the 1-based escape iteration for one pixel, or -1.
        /// </summary>
        public static int PerturbZ2(
            double[] orbitRe,
            double[] orbitIm,
            double[] rebaseRe,
            double[] rebaseIm,
            double dz0Re,
            double dz0Im,
            double dcRe,
            double dcIm,
            int maxIter,
            double escapeRadius,
            out double finalRe,
            out double finalIm)
        {
            double dzr = dz0Re, dzi = dz0Im;
            double[] refRe = orbitRe, refIm = orbitIm;
            int m = 0;
            int last = refRe.Length - 1;
            double r2 = escapeRadius * escapeRadius;
            for (int it = 1; it <= maxIter; it++)
            {
                // dz = (2*Z[m] + dz) * dz + dc, with the reference's fma-contracted multiply
                double tr = 2.0 * refRe[m] + dzr;
                double ti = 2.0 * refIm[m] + dzi;
                var (mr, mi) = FractalEngine.ComplexMul(tr, ti, dzr, dzi);
                dzr = mr + dcRe;
                dzi = mi + dcIm;
                m = Math.Min(m + 1, last);
                double zr = refRe[m] + dzr;
                double zi = refIm[m] + dzi;
                double zabs2 = zr * zr + zi * zi;
                if (zabs2 > r2)
                {
                    finalRe = zr;
                    finalIm = zi;
                    return it;
                }
                if (zabs2 < dzr * dzr + dzi * dzi)
                {
                    dzr = zr;
                    dzi = zi;
                    refRe = rebaseRe;
                    refIm = rebaseIm;
                    last = refRe.Length - 1;
                    m = 0;
                }
            }
            finalRe = 0.0;
            finalIm = 0.0;
            return -1;
        }

        /// <summary>
        /// <see cref="PerturbZ2(double[],double[],double[],double[],double,double,double,double,int,double,out double,out double)"/>
        /// that also carries the derivative of the full z for a distance estimate
        /// (spec/fractals.md "Distance estimate"): updated from the z the previous iteration
        /// reconstructed — fl(Z_0 + delta_0) before the first — and left alone by a rebase,
        /// which changes delta, m and the orbit followed but not z. Mandelbrot: d0 = (0, 0),
        /// <paramref name="addScale"/>; Julia: d0 = (ps, 0), no addition. Counts and the final
        /// z are exactly PerturbZ2's.
        /// </summary>
        internal static int PerturbZ2De(
            double[] orbitRe,
            double[] orbitIm,
            double[] rebaseRe,
            double[] rebaseIm,
            double dz0Re,
            double dz0Im,
            double dcRe,
            double dcIm,
            int maxIter,
            double escapeRadius,
            double d0Re,
            double d0Im,
            bool addScale,
            double ps,
            out double finalRe,
            out double finalIm,
            out double finalDr,
            out double finalDi)
        {
            double dzr = dz0Re, dzi = dz0Im;
            double[] refRe = orbitRe, refIm = orbitIm;
            int m = 0;
            int last = refRe.Length - 1;
            double r2 = escapeRadius * escapeRadius;
            double dr = d0Re, di = d0Im;
            double zr = refRe[0] + dzr;                   // the pre-square z of iteration 1
            double zi = refIm[0] + dzi;
            for (int it = 1; it <= maxIter; it++)
            {
                double tdr = 2.0 * (zr * dr - zi * di);
                if (addScale)
                    tdr = tdr + ps;
                double tdi = 2.0 * (zr * di + zi * dr);
                dr = tdr;
                di = tdi;
                double tr = 2.0 * refRe[m] + dzr;
                double ti = 2.0 * refIm[m] + dzi;
                var (mr, mi) = FractalEngine.ComplexMul(tr, ti, dzr, dzi);
                dzr = mr + dcRe;
                dzi = mi + dcIm;
                m = Math.Min(m + 1, last);
                zr = refRe[m] + dzr;
                zi = refIm[m] + dzi;
                double zabs2 = zr * zr + zi * zi;
                if (zabs2 > r2)
                {
                    finalRe = zr;
                    finalIm = zi;
                    finalDr = dr;
                    finalDi = di;
                    return it;
                }
                if (zabs2 < dzr * dzr + dzi * dzi)
                {
                    // z, and so the derivative, are unchanged: only delta, m and the orbit move
                    dzr = zr;
                    dzi = zi;
                    refRe = rebaseRe;
                    refIm = rebaseIm;
                    last = refRe.Length - 1;
                    m = 0;
                }
            }
            finalRe = finalIm = finalDr = finalDi = 0.0;
            return -1;
        }

        /// <summary>
        /// Mandelbrot perturbation with BLA skips (spec/deep-zoom.md "BLA"): each pass, a
        /// pixel at a stride-aligned reference index takes the longest live span of
        /// <paramref name="table"/> whose radius exceeds |dz| and that fits in maxIter, else
        /// one plain step exactly as <see cref="PerturbZ2(double[],double[],double,double,double,double,int,double,out double,out double)"/>
        /// takes it. A skip is plain doubles: dz' = A·dz + B·dc, and with
        /// <paramref name="distance"/> the derivative d' = A·d + B·ps. Returns the count
        /// (-1 if none) and how many skips the pixel took.
        /// </summary>
        internal static int PerturbZ2Bla(
            double[] orbitRe,
            double[] orbitIm,
            BlaTable table,
            double dcRe,
            double dcIm,
            int maxIter,
            double escapeRadius,
            bool distance,
            double ps,
            out double finalRe,
            out double finalIm,
            out double finalDr,
            out double finalDi,
            out int applications)
        {
            const int stride = BlaTable.Stride;
            double dzr = 0.0, dzi = 0.0;
            int m = 0;
            long n = 0;
            int last = orbitRe.Length - 1;
            double r2 = escapeRadius * escapeRadius;
            double dr = 0.0, di = 0.0;
            double zr = orbitRe[0] + dzr;                  // the pre-square z of the first pass
            double zi = orbitIm[0] + dzi;
            int levels = table.Levels;
            int blocks = levels > 0 ? table.R[0].Length : 0;
            applications = 0;
            while (n < maxIter)
            {
                int chosen = -1;
                if (levels > 0 && m % stride == 0 && m / stride < blocks)
                {
                    double dm = BlaTable.Mag(dzr, dzi);
                    for (int level = 0; level < levels; level++)
                    {
                        long span = (long)stride << level;
                        if (m % span != 0 || m / span >= table.R[level].Length || n + span > maxIter
                            || !(dm < table.R[level][m / span]))
                            break;
                        chosen = level;
                    }
                }
                if (chosen >= 0)
                {
                    int span = stride << chosen;
                    int e = m / span;
                    double ar = table.Ar[chosen][e], ai = table.Ai[chosen][e];
                    double br = table.Br[chosen][e], bi = table.Bi[chosen][e];
                    if (distance)
                    {
                        double tdr = (ar * dr - ai * di) + br * ps;
                        double tdi = (ar * di + ai * dr) + bi * ps;
                        dr = tdr;
                        di = tdi;
                    }
                    double nr = (ar * dzr - ai * dzi) + (br * dcRe - bi * dcIm);
                    double ni = (ar * dzi + ai * dzr) + (br * dcIm + bi * dcRe);
                    dzr = nr;
                    dzi = ni;
                    m += span;
                    n += span;
                    applications++;
                }
                else
                {
                    if (distance)
                    {
                        double tdr = 2.0 * (zr * dr - zi * di) + ps;
                        double tdi = 2.0 * (zr * di + zi * dr);
                        dr = tdr;
                        di = tdi;
                    }
                    double tr = 2.0 * orbitRe[m] + dzr;
                    double ti = 2.0 * orbitIm[m] + dzi;
                    var (mr, mi) = FractalEngine.ComplexMul(tr, ti, dzr, dzi);
                    dzr = mr + dcRe;
                    dzi = mi + dcIm;
                    m = Math.Min(m + 1, last);
                    n++;
                }
                zr = orbitRe[m] + dzr;
                zi = orbitIm[m] + dzi;
                double zabs2 = zr * zr + zi * zi;
                if (zabs2 > r2)
                {
                    finalRe = zr;
                    finalIm = zi;
                    finalDr = dr;
                    finalDi = di;
                    return (int)n;
                }
                if (zabs2 < dzr * dzr + dzi * dzi)
                {
                    dzr = zr;                               // z and the derivative are unchanged
                    dzi = zi;
                    m = 0;
                }
            }
            finalRe = finalIm = finalDr = finalDi = 0.0;
            return -1;
        }

        /// <summary>Component-form perturbation for the Burning Ship, using stable diffabs.</summary>
        public static int PerturbBurningShip(
            double[] orbitRe,
            double[] orbitIm,
            double dcRe,
            double dcIm,
            int maxIter,
            double escapeRadius,
            out double finalRe,
            out double finalIm)
        {
            double dx = 0.0, dy = 0.0;
            int m = 0;
            int last = orbitRe.Length - 1;
            double r2 = escapeRadius * escapeRadius;
            for (int it = 1; it <= maxIter; it++)
            {
                double xRef = orbitRe[m];
                double yRef = orbitIm[m];
                double a = DiffAbs(xRef, dx);
                double b = DiffAbs(yRef, dy);
                double newDx = (2.0 * xRef + dx) * dx - (2.0 * yRef + dy) * dy + dcRe;
                double newDy = 2.0 * (Math.Abs(xRef) * b + Math.Abs(yRef) * a + a * b) + dcIm;
                dx = newDx;
                dy = newDy;
                m = Math.Min(m + 1, last);
                double zx = orbitRe[m] + dx;
                double zy = orbitIm[m] + dy;
                double zabs2 = zx * zx + zy * zy;
                if (zabs2 > r2)
                {
                    finalRe = zx;
                    finalIm = zy;
                    return it;
                }
                if (zabs2 < dx * dx + dy * dy)
                {
                    dx = zx;
                    dy = zy;
                    m = 0;
                }
            }
            finalRe = 0.0;
            finalIm = 0.0;
            return -1;
        }

        /// <summary>|ref + delta| - |ref|, computed without cancellation (case analysis).</summary>
        private static double DiffAbs(double reference, double delta)
        {
            double total = reference + delta;
            if (reference >= 0.0)
                return total >= 0.0 ? delta : -(2.0 * reference + delta);
            return total <= 0.0 ? -delta : 2.0 * reference + delta;
        }
    }
}
